pkg/gameservers/gameservers.go

```diff
@@ -34,19 +35,37 @@ func isGameServerPod(pod *corev1.Pod) bool {
 	return false
 }
 
-// address returns the IP that the given Pod is run on.
-// This will default to the ExternalIP, but if the ExternalIP is
-// not set, it will fall back to the InternalIP,
+// address returns the network address that the given Pod is run on.
+// This will default to the ExternalDNS, but if the ExternalDNS is
+// not set, it will fall back to the ExternalIP then InternalDNS then InternalIP,
+// If externalDNS is false, skip ExternalDNS and InternalDNS.
 // since we can have clusters that are private, and/or tools like minikube
 // that only report an InternalIP.
-func address(node *corev1.Node) (string, error) {
+func address(node *corev1.Node, externalDNS bool) (string, error) {
+
+	if externalDNS {
+		for _, a := range node.Status.Addresses {
+			if a.Type == corev1.NodeExternalDNS {
+				return a.Address, nil
+			}
+		}
+	}
+
 	for _, a := range node.Status.Addresses {
 		if a.Type == corev1.NodeExternalIP && net.ParseIP(a.Address) != nil {
 			return a.Address, nil
 		}
 	}
 
-	// There might not be a public IP, so fall back to the private IP
+	// There might not be a public DNS/IP, so fall back to the private DNS/IP
+	if externalDNS {
+		for _, a := range node.Status.Addresses {
+			if a.Type == corev1.NodeInternalDNS {
+				return a.Address, nil
+			}
+		}
+	}
+
```

Comment:
Why this is a duplicate to lines 46-53?

Reply:
The code is not duplicated.
Lines 46-53 are a search for External DNS.
This code is a search for Internal DNS.

Did you point out that you should search both External DNS and Internal DNS at once?

Reply:
Ok, now I see the difference. Note that there is a similar code down the file:
```
if a.Type == corev1.NodeInternalIP && net.ParseIP(a.Address) != nil {
```
So probably this code is redundant. WDYT? The only difference is `net.ParseIP` call.

Reply:
To be honest, I was quite confused. If I were to implement it like a table-based test, it would look like this.

```
func address(node *corev1.Node) (string, error) {
	var targetTypes []corev1.NodeAddressType
	if runtime.FeatureEnabled(runtime.NodeExternalDNS) {
		targetTypes = []corev1.NodeAddressType{
			corev1.NodeExternalDNS, corev1.NodeExternalIP, corev1.NodeInternalDNS, corev1.NodeInternalIP,
		}
	} else {
		targetTypes = []corev1.NodeAddressType{corev1.NodeExternalIP, corev1.NodeInternalIP}
	}

	for _, addressType := range targetTypes {
		for _, a := range node.Status.Addresses {
			if a.Type != addressType {
				continue
			}
			if addressType == corev1.NodeExternalIP || addressType == corev1.NodeInternalIP {
				if net.ParseIP(a.Address) != nil {
					return a.Address, nil
				}
			} else {
				return a.Address, nil
			}
		}
	}

	return "", errors.Errorf("Could not find an address for Node: %s", node.ObjectMeta.Name)
}
```

After writing this far, I didn't think this code was smart. The current code may be silly and dumb, but it is terribly simple and easy to read.
If the amount of code increases further, it should be revised, but for the current amount, I thought the simplicity was more valuable.

Reply:
That's right, thanks